### PR TITLE
Add CSM NEG support.

### PR DESCRIPTION
**Pull request by cadmuxe**

### CODE REVIEW

**Comment by cadmuxe** [comment]
This is a draft, need add tests.

File: cmd/glbc/main.go [hunk 1]
```diff
@@ -197,7 +205,7 @@ func runControllers(ctx *ingctx.ControllerContext) {
 	fwc := firewalls.NewFirewallController(ctx, flags.F.NodePortRanges.Values())
 
 	// TODO: Refactor NEG to use cloud mocks so ctx.Cloud can be referenced within NewController.
-	negController := neg.NewController(negtypes.NewAdapter(ctx.Cloud), ctx, lbc.Translator, ctx.ClusterNamer, flags.F.ResyncPeriod, flags.F.NegGCPeriod, neg.NegSyncerType(flags.F.NegSyncerType), flags.F.EnableReadinessReflector)
+	negController := neg.NewController(negtypes.NewAdapter(ctx.Cloud), ctx, lbc.Translator, ctx.ClusterNamer, flags.F.ResyncPeriod, flags.F.NegGCPeriod, neg.NegSyncerType(flags.F.NegSyncerType), flags.F.EnableReadinessReflector, flags.F.EnableCSM)
```
Comment (freehan) [inline cmd/glbc/main.go]: refactor neg.NewController to take controller context

File: pkg/annotations/destination_rule.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package annotations
+
+import "encoding/json"
+
+// PortSubsetNegMap is the mapping between service port:subset to NEG name
+type PortSubsetNegMap map[string]string
```
Comment (freehan) [inline pkg/annotations/destination_rule.go]: you probably need `map[string]map[string]string`?

```

"cloud.google.com/neg-status": `{         
  "network_endpoint_groups": {              
       "v1": {                 
                    "9080": "somehash-default-reviews-v1-9080",               
         }              
        "v2": {
                   "9080": "somehash-default-reviews-v2-9080",               
          }
}




```

File: pkg/utils/namer.go [hunk 3]
```diff
@@ -422,12 +422,32 @@ func (n *Namer) NamedPort(port int64) string {
 // of all NEGs associated with the current cluster. Any modifications
 // must be backward compatible.
 func (n *Namer) NEG(namespace, name string, port int32) string {
+	return n.NEGWithSubset(namespace, name, "", port)
+}
+
+// NEGWithSubset returns the gce neg name based on the service namespace, name
+// target port and Istio:DestinationRule subset if provided. NEG naming convention:
+//
+//   {prefix}{version}-{clusterid}-{namespace}-{name}-{service port}[-subset]-{hash}
```
Comment (freehan) [inline pkg/utils/namer.go]: {prefix}{version}-{clusterid}-{namespace}-{name}-{subset}-{service port}-{hash}?

Comment (freehan) [reply pkg/utils/namer.go]: consider spliting the namer NegForServicePort
NegForDestinationRule

File: cmd/glbc/main.go (hunk 1 above)
Comment (cadmuxe) [reply cmd/glbc/main.go]: Discussed offline, will keep this by now.

File: pkg/annotations/destination_rule.go (hunk 2 above)
Comment (cadmuxe) [reply pkg/annotations/destination_rule.go]: Updated.

File: pkg/utils/namer.go (hunk 3 above)
Comment (cadmuxe) [reply pkg/utils/namer.go]: Updated.

File: pkg/neg/controller.go [hunk 4]
```diff
@@ -316,6 +348,22 @@ func (c *Controller) processService(key string) error {
 		}
 	}
 
+	if c.enableCSM {
+		// Find all destination rules that using this service.
+		destinationRules := getDestinationRulesFromStore(c.destinationRuleLister, service)
+		servicePorts := gatherPortMappingFromService(service)
```
Comment (freehan) [inline pkg/neg/controller.go]: IIUC, this only creates NEGs per each (destination rule subsets X service ports)

If c.enableCSM, you also need to create NEGs for each service port. Except for kube-system services and `kubernetes.default` service

File: pkg/neg/controller.go [hunk 5]
```diff
@@ -316,6 +348,22 @@ func (c *Controller) processService(key string) error {
...
+		for destinationRuleNN, destinationRule := range destinationRules {
+			destinationRulePortInfoMap := negtypes.NewPortInfoMapWithDestinationRule(namespace, name, servicePorts, c.namer, true, destinationRule)
```
Comment (freehan) [inline pkg/neg/controller.go]: Need to validate if the destinationRule has unique subset names. If not, error out. 

I confirmed with the Istio team that they want user to create unique subset names but did not enforce it with validation.

File: pkg/neg/controller.go [hunk 6]
```diff
@@ -422,6 +472,44 @@ func (c *Controller) syncNegStatusAnnotation(namespace, name string, portMap neg
 	return err
 }
 
+func (c *Controller) syncDestinationRuleNegStatusAnnotation(namespace, destinationRuleName string, portmap negtypes.PortInfoMap) error {
```
Comment (freehan) [inline pkg/neg/controller.go]: add comment

File: pkg/neg/controller.go [hunk 7]
```diff
@@ -463,6 +551,30 @@ func (c *Controller) enqueueIngressServices(ing *v1beta1.Ingress) {
 	}
 }
 
+func (c *Controller) enqueueDestinationRule(obj interface{}) {
```
Comment (freehan) [inline pkg/neg/controller.go]: add a comment

File: pkg/neg/types/types.go [hunk 8]
```diff
@@ -126,10 +179,17 @@ type NegSyncerKey struct {
 	Port int32
 	// Service target port
 	TargetPort string
+
+	// Subset name, subset is defined in Istio:DestinationRule, this is only used
+	// when --enable-csm=true.
+	Subset string
+
+	// Subset label, should set together with Subset.
+	SubsetLabels string
```
[the remark this answers is not in the document]
Comment (freehan) [reply pkg/neg/types/types.go]: NegSyncerKey is to uniquely identify a NegSyncer. A NegSyncer maps to one NEG per zone

File: pkg/neg/types/types.go [hunk 9]
```diff
@@ -34,6 +36,14 @@ type PortInfo struct {
 	// TargetPort is the target port of the service port
 	// This can be port number or a named port
 	TargetPort string
+
+	// Subset name, subset is defined in Istio:DestinationRule, this is only used
+	// when --enable-csm=true.
+	Subset string
+
+	// Subset label, should set together with Subset.
+	SubsetLabels string
```
[the remark this answers is not in the document]
Comment (freehan) [reply pkg/neg/types/types.go]: okay. Got it. If it is simpler, then just do it this way.

**Pull request merged**